File: src/oauthkitchen/analyzers/shadow.py
"""Shadow OAuth Detector - identifies risky OAuth exposure patterns."""

from __future__ import annotations

from oauthkitchen.analyzers.translator import PermissionTranslator
from oauthkitchen.config import Config
from oauthkitchen.models import (
    AnalysisResult,
    AppType,
    ConsentType,
    RiskCategory,
    ServicePrincipal,
    ShadowOAuthFinding,
)
from oauthkitchen.utils.logging import get_logger

logger = get_logger("shadow")

# High-impact scopes that warrant extra scrutiny
HIGH_IMPACT_SCOPES = {
    "Directory.ReadWrite.All",
    "RoleManagement.ReadWrite.Directory",
    "Application.ReadWrite.All",
    "AppRoleAssignment.ReadWrite.All",
    "Mail.ReadWrite",
    "Mail.Read",
    "Files.ReadWrite.All",
    "Files.Read.All",
    "Sites.ReadWrite.All",
    "User.ReadWrite.All",
    "Group.ReadWrite.All",
    "GroupMember.ReadWrite.All",
    "Chat.Read.All",
    "ChannelMessage.Read.All",
}


class ShadowOAuthDetector:
    """
    Detects shadow OAuth patterns that indicate potential security risks.

    Shadow OAuth patterns include:
    - External apps with delegated grants not owned by tenant
    - User-consented apps with high-impact scopes
    - Apps with offline_access (long-lived refresh tokens)
    - Inactive apps with elevated permissions
    - Apps without owners but with high privileges
    """

    def __init__(
        self,
        config: Config,
        translator: PermissionTranslator | None = None
    ):
        """
        Initialize the detector.

        Args:
            config: Configuration object
            translator: Permission translator for impact assessment
        """
        self.config = config
        self.thresholds = config.thresholds
        self.translator = translator or PermissionTranslator()

    def detect(
        self,
        result: AnalysisResult
    ) -> list[ShadowOAuthFinding]:
        """
        Run all shadow OAuth detection rules.

        Args:
            result: Analysis result with collected data

        Returns:
            List of findings
        """
        logger.info("Running shadow OAuth detection...")

        findings: list[ShadowOAuthFinding] = []

        for sp in result.service_principals:
            # Skip first-party Microsoft apps
            if sp.app_type == AppType.FIRST_PARTY_MICROSOFT:
                continue

            # Skip apps on allow list
            if sp.app_id in self.config.allow_deny.allowed_app_ids:
                continue

            # Run detection rules
            findings.extend(self._detect_external_delegated_grants(sp))
            findings.extend(self._detect_user_consent_high_impact(sp))
            findings.extend(self._detect_offline_access_risk(sp))
            findings.extend(self._detect_inactive_privileged(sp, result))
            findings.extend(self._detect_orphaned_privileged(sp))
            findings.extend(self._detect_unknown_publisher_high_impact(sp))

        # Sort by severity
        severity_order = {"Critical": 0, "High": 1, "Medium": 2, "Low": 3}
        findings.sort(key=lambda f: severity_order.get(f.severity, 99))

        result.shadow_findings = findings

        logger.info("Found %d shadow OAuth findings", len(findings))
        return findings

    def _detect_external_delegated_grants(
        self,
        sp: ServicePrincipal
    ) -> list[ShadowOAuthFinding]:
        """Detect external apps with delegated permission grants."""
        findings: list[ShadowOAuthFinding] = []

        # Only flag external/multi-tenant apps
        if sp.app_type not in [AppType.THIRD_PARTY_MULTI_TENANT, AppType.EXTERNAL_UNKNOWN]:
            return findings

        if not sp.has_delegated_grants:
            return findings

        # Check for high-impact scopes
        high_impact = sp.all_delegated_scopes & HIGH_IMPACT_SCOPES
        all_scopes = list(sp.all_delegated_scopes)

        if high_impact:
            findings.append(ShadowOAuthFinding(
                finding_type="external_delegated_high_impact",
                severity="High",
                title="External App with High-Impact Delegated Permissions",
                description=(
                    f"Third-party app '{sp.display_name}' has delegated permissions "
                    f"including high-impact scopes. This app is not owned by your tenant."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(high_impact),
                affected_user_count=sp.consent_user_count,
                recommendation=(
                    "Review if this app is still needed. Verify the publisher and "
                    "consider reducing permissions or removing the app."
                )
            ))
        elif all_scopes:
            findings.append(ShadowOAuthFinding(
                finding_type="external_delegated",
                severity="Medium",
                title="External App with Delegated Permissions",
                description=(
                    f"Third-party app '{sp.display_name}' has delegated permissions. "
                    f"This app is not owned by your tenant."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=all_scopes,
                affected_user_count=sp.consent_user_count,
                recommendation=(
                    "Verify this app is authorized for use in your organization."
                )
            ))

        return findings

    def _detect_user_consent_high_impact(
        self,
        sp: ServicePrincipal
    ) -> list[ShadowOAuthFinding]:
        """Detect user-consented apps with high-impact scopes."""
        findings: list[ShadowOAuthFinding] = []

        # Check for user consent
        user_consented_scopes: set[str] = set()
        for grant in sp.oauth2_permission_grants:
            if grant.consent_type == ConsentType.USER:
                user_consented_scopes.update(grant.scopes)

        if not user_consented_scopes:
            return findings

        # Check for high-impact scopes via user consent
        high_impact = user_consented_scopes & HIGH_IMPACT_SCOPES

        if high_impact:
            findings.append(ShadowOAuthFinding(
                finding_type="user_consent_high_impact",
                severity="High",
                title="User-Consented App with High-Impact Permissions",
                description=(
                    f"App '{sp.display_name}' received high-impact permissions via user consent "
                    f"(not admin consent). {sp.consent_user_count} user(s) granted consent."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(high_impact),
                affected_user_count=sp.consent_user_count,
                recommendation=(
                    "Review user consent settings. Consider blocking user consent for "
                    "high-impact permissions and requiring admin approval."
                )
            ))

        return findings

    def _detect_offline_access_risk(
        self,
        sp: ServicePrincipal
    ) -> list[ShadowOAuthFinding]:
        """Detect apps with offline_access that increase token persistence risk."""
        findings: list[ShadowOAuthFinding] = []

        if "offline_access" not in sp.all_delegated_scopes:
            return findings

        # Check if combined with high-impact scopes
        other_scopes = sp.all_delegated_scopes - {"offline_access", "openid", "profile", "email"}
        high_impact = other_scopes & HIGH_IMPACT_SCOPES

        if high_impact:
            findings.append(ShadowOAuthFinding(
                finding_type="offline_access_high_impact",
                severity="Medium",
                title="Long-Lived Token Risk with High-Impact Permissions",
                description=(
                    f"App '{sp.display_name}' has offline_access (refresh tokens) combined "
                    f"with high-impact permissions. This allows persistent access even after "
                    f"password changes."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(high_impact | {"offline_access"}),
                affected_user_count=sp.consent_user_count,
                recommendation=(
                    "Consider if offline_access is necessary. Review token lifetime policies "
                    "and implement Conditional Access for app access controls."
                )
            ))

        return findings

    def _detect_inactive_privileged(
        self,
        sp: ServicePrincipal,
        result: AnalysisResult
    ) -> list[ShadowOAuthFinding]:
        """Detect inactive apps with elevated permissions."""
        findings: list[ShadowOAuthFinding] = []

        # Skip if sign-in data not available
        if not result.sign_in_data_available:
            return findings

        if not sp.sign_in_activity or not sp.sign_in_activity.data_available:
            return findings

        days_inactive = sp.sign_in_activity.days_since_last_activity
        threshold = self.thresholds.inactive_days_threshold

        # Check if inactive
        if days_inactive is None:
            # Never used
            is_inactive = True
            inactive_msg = "never been used"
        elif days_inactive > threshold:
            is_inactive = True
            inactive_msg = f"not been used in {days_inactive} days"
        else:
            is_inactive = False

        if not is_inactive:
            return findings

        # Check for high-impact permissions
        all_perms = sp.all_delegated_scopes | sp.all_app_role_values
        high_impact = all_perms & HIGH_IMPACT_SCOPES

        if high_impact:
            findings.append(ShadowOAuthFinding(
                finding_type="inactive_privileged",
                severity="High",
                title="Inactive App with High-Impact Permissions",
                description=(
                    f"App '{sp.display_name}' has {inactive_msg} but retains "
                    f"high-impact permissions."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(high_impact),
                recommendation=(
                    f"Review if this app is still needed. Consider removing or reducing "
                    f"permissions for unused apps (inactive threshold: {threshold} days)."
                )
            ))

        return findings

    def _detect_orphaned_privileged(
        self,
        sp: ServicePrincipal
    ) -> list[ShadowOAuthFinding]:
        """Detect apps without owners that have elevated permissions."""
        findings: list[ShadowOAuthFinding] = []

        if sp.has_owners:
            return findings

        # Check for high-impact permissions
        all_perms = sp.all_delegated_scopes | sp.all_app_role_values
        high_impact = all_perms & HIGH_IMPACT_SCOPES

        if high_impact:
            findings.append(ShadowOAuthFinding(
                finding_type="orphaned_privileged",
                severity="High",
                title="Orphaned App with High-Impact Permissions",
                description=(
                    f"App '{sp.display_name}' has no assigned owners but has "
                    f"high-impact permissions."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(high_impact),
                recommendation=(
                    "Assign owners to this app to ensure accountability. "
                    "Consider if the permissions are still needed."
                )
            ))
        elif all_perms:
            findings.append(ShadowOAuthFinding(
                finding_type="orphaned_with_permissions",
                severity="Medium",
                title="Orphaned App with Permissions",
                description=(
                    f"App '{sp.display_name}' has no assigned owners but has "
                    f"active permissions."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(all_perms),
                recommendation="Assign owners to this app to ensure accountability."
            ))

        return findings

    def _detect_unknown_publisher_high_impact(
        self,
        sp: ServicePrincipal
    ) -> list[ShadowOAuthFinding]:
        """Detect apps without verified publisher that have high-impact permissions."""
        findings: list[ShadowOAuthFinding] = []

        if sp.has_verified_publisher:
            return findings

        # Skip tenant-owned apps (we control them)
        if sp.app_type == AppType.TENANT_OWNED:
            return findings

        # Check for high-impact permissions
        all_perms = sp.all_delegated_scopes | sp.all_app_role_values
        high_impact = all_perms & HIGH_IMPACT_SCOPES

        if high_impact:
            findings.append(ShadowOAuthFinding(
                finding_type="unverified_publisher_high_impact",
                severity="High",
                title="Unverified Publisher with High-Impact Permissions",
                description=(
                    f"App '{sp.display_name}' has no verified publisher but has "
                    f"high-impact permissions."
                ),
                service_principal_id=sp.object_id,
                service_principal_name=sp.display_name,
                affected_scopes=list(high_impact),
                recommendation=(
                    "Verify the legitimacy of this app. Consider requiring verified "
                    "publishers for apps with high-impact permissions."
                )
            ))

        return findings

    def get_safe_to_review_first(
        self,
        findings: list[ShadowOAuthFinding]
    ) -> list[ShadowOAuthFinding]:
        """
        Get findings that are safe to review first.

        These are typically lower risk to investigate and may be quick wins.
        """
        safe_types = {
            "orphaned_with_permissions",
            "inactive_privileged",
            "orphaned_privileged",
        }

        return [
            f for f in findings
            if f.finding_type in safe_types
        ]

    def get_blast_radius_report(
        self,
        findings: list[ShadowOAuthFinding],
        service_principals: list[ServicePrincipal]
    ) -> list[dict]:
        """
        Generate blast radius hints showing impact scope.

        Returns list of dicts with:
        - sp_id, sp_name
        - affected_users: count
        - resource_apis: list of affected APIs
        - risk_summary
        """
        sp_map = {sp.object_id: sp for sp in service_principals}
        results = []

        seen_sps: set[str] = set()

        for finding in findings:
            if finding.service_principal_id in seen_sps:
                continue
            seen_sps.add(finding.service_principal_id)

            sp = sp_map.get(finding.service_principal_id)
            if not sp:
                continue

            # Determine affected resource APIs
            resource_apis: set[str] = set()
            for grant in sp.oauth2_permission_grants:
                # Resource ID maps to a service principal
                if grant.resource_id in sp_map:
                    resource_apis.add(sp_map[grant.resource_id].display_name)
                else:
                    resource_apis.add(f"Resource: {grant.resource_id[:8]}...")

            for assignment in sp.app_role_assignments:
                if assignment.resource_display_name:
                    resource_apis.add(assignment.resource_display_name)
                elif assignment.resource_id in sp_map:
                    resource_apis.add(sp_map[assignment.resource_id].display_name)

            results.append({
                "sp_id": sp.object_id,
                "sp_name": sp.display_name,
                "app_id": sp.app_id,
                "affected_users": sp.consent_user_count,
                "resource_apis": sorted(resource_apis),
                "total_permissions": len(sp.all_delegated_scopes) + len(sp.all_app_role_values),
                "severity": finding.severity,
            })

        # Sort by affected users descending
        results.sort(key=lambda x: x["affected_users"], reverse=True)

        return results

File: web/frontend/src/lib/analyzers/shadow.ts
/**
 * Shadow OAuth Detector — finds risky OAuth exposure patterns.
 *
 * Ported from src/oauthkitchen/analyzers/shadow.py
 */

import {
  type ServicePrincipal,
  type ShadowOAuthFinding,
  AppType,
  ConsentType,
  getAllDelegatedScopes,
  getAllAppRoleValues,
  spHasVerifiedPublisher,
  hasOwners,
  daysSinceLastActivity,
} from '@/types/models'
import { permissionTranslator } from '@/lib/analyzers/translator'

// ============================================================================
// HIGH-IMPACT SCOPE LIST
// ============================================================================

const HIGH_IMPACT_SCOPES = new Set([
  'Directory.ReadWrite.All',
  'RoleManagement.ReadWrite.Directory',
  'Application.ReadWrite.All',
  'AppRoleAssignment.ReadWrite.All',
  'DelegatedPermissionGrant.ReadWrite.All',
  'GroupMember.ReadWrite.All',
  'Group.ReadWrite.All',
  'Mail.ReadWrite',
  'Mail.Read',
  'Files.ReadWrite.All',
  'Sites.ReadWrite.All',
  'Sites.FullControl.All',
  'User.ReadWrite.All',
  'User.Read.All',
  'AuditLog.Read.All',
  'Policy.ReadWrite.All',
  'PrivilegedAccess.ReadWrite.AzureAD',
  'RoleAssignmentSchedule.ReadWrite.Directory',
  'EntitlementManagement.ReadWrite.All',
  'Chat.Read.All',
  'ChannelMessage.Read.All',
  'Team.ReadWrite.All',
  'DeviceManagementApps.ReadWrite.All',
  'DeviceManagementConfiguration.ReadWrite.All',
  'SecurityEvents.ReadWrite.All',
  'SecurityActions.ReadWrite.All',
  'ThreatIndicators.ReadWrite.OwnedBy',
  'AccessReview.ReadWrite.All',
  'ProgramControl.ReadWrite.All',
  'Reports.Read.All',
  'Calendars.ReadWrite',
  'Contacts.ReadWrite',
  'Tasks.ReadWrite',
  'MailboxSettings.ReadWrite',
  'OnlineMeetingRecording.Read.All',
])

// Score threshold above which a permission is considered high-impact
const HIGH_IMPACT_SCORE_THRESHOLD = 65

// ============================================================================
// HELPER
// ============================================================================

function isHighImpact(scope: string): boolean {
  // Check hardcoded list first, then fall back to translator score
  if (HIGH_IMPACT_SCOPES.has(scope)) return true
  const translated = permissionTranslator.translate(scope)
  return translated.impactScore >= HIGH_IMPACT_SCORE_THRESHOLD
}

function isExternal(sp: ServicePrincipal): boolean {
  return (
    sp.appType === AppType.THIRD_PARTY_MULTI_TENANT || sp.appType === AppType.EXTERNAL_UNKNOWN
  )
}

// ============================================================================
// SHADOW OAUTH DETECTOR
// ============================================================================

export class ShadowOAuthDetector {
  private includeRemediation: boolean
  private inactiveThresholdDays: number

  constructor(includeRemediation = false, inactiveThresholdDays = 90) {
    this.includeRemediation = includeRemediation
    this.inactiveThresholdDays = inactiveThresholdDays
  }

  detect(principals: ServicePrincipal[]): ShadowOAuthFinding[] {
    const findings: ShadowOAuthFinding[] = []

    for (const sp of principals) {
      // Skip Microsoft first-party apps
      if (sp.appType === AppType.FIRST_PARTY_MICROSOFT) continue

      findings.push(...this._detectExternalDelegatedHighImpact(sp))
      findings.push(...this._detectUserConsentHighImpact(sp))
      findings.push(...this._detectOfflineAccessRisk(sp))
      findings.push(...this._detectInactivePrivileged(sp))
      findings.push(...this._detectOrphanedPrivileged(sp))
      findings.push(...this._detectUnverifiedPublisherHighImpact(sp))
    }

    return findings
  }

  // --------------------------------------------------------------------------
  // DETECTION PATTERNS
  // --------------------------------------------------------------------------

  private _detectExternalDelegatedHighImpact(sp: ServicePrincipal): ShadowOAuthFinding[] {
    if (!isExternal(sp)) return []

    const delegatedScopes = getAllDelegatedScopes(sp)
    const highImpactFound = Array.from(delegatedScopes).filter(isHighImpact)
    if (highImpactFound.length === 0) return []

    const userCount =
      new Set(
        (sp.oauth2PermissionGrants || [])
          .filter((g) => g.principalId)
          .map((g) => g.principalId as string)
      ).size

    return [
      {
        findingType: 'external_delegated_high_impact',
        severity: 'critical',
        title: 'External app with high-impact delegated permissions',
        description:
          `"${sp.displayName}" is an external/third-party app with ${highImpactFound.length} ` +
          `high-impact delegated permission(s) granted by ${userCount} user(s). ` +
          `This creates an exfiltration risk for data accessible to those users.`,
        servicePrincipalId: sp.objectId,
        servicePrincipalName: sp.displayName,
        affectedScopes: highImpactFound,
        affectedUserCount: userCount,
        recommendation: this.includeRemediation
          ? 'Review whether this app needs these permissions. Consider restricting via Conditional Access or revoking user consent.'
          : null,
      },
    ]
  }

  private _detectUserConsentHighImpact(sp: ServicePrincipal): ShadowOAuthFinding[] {
    const userConsentGrants = (sp.oauth2PermissionGrants || []).filter(
      (g) => g.consentType === ConsentType.USER && g.principalId
    )
    if (userConsentGrants.length === 0) return []

    const userConsentScopes = new Set<string>()
    for (const grant of userConsentGrants) {
      grant.scope
        .split(/\s+/)
        .filter((s) => s.trim())
        .forEach((s) => userConsentScopes.add(s))
    }

    const highImpactFound = Array.from(userConsentScopes).filter(isHighImpact)
    if (highImpactFound.length === 0) return []

    const userCount = new Set(userConsentGrants.map((g) => g.principalId as string)).size

    return [
      {
        findingType: 'user_consent_high_impact',
        severity: 'high',
        title: 'User-consented high-impact permissions',
        description:
          `"${sp.displayName}" has high-impact permissions granted via individual user consent ` +
          `(${userCount} user(s)). These were not reviewed by an administrator and may violate ` +
          `security policies.`,
        servicePrincipalId: sp.objectId,
        servicePrincipalName: sp.displayName,
        affectedScopes: highImpactFound,
        affectedUserCount: userCount,
        recommendation: this.includeRemediation
          ? 'Revoke user-consented grants and replace with admin-consented grants after review. Enable admin consent requirement in Azure AD.'
          : null,
      },
    ]
  }

  private _detectOfflineAccessRisk(sp: ServicePrincipal): ShadowOAuthFinding[] {
    const allScopes = getAllDelegatedScopes(sp)
    if (!allScopes.has('offline_access') && !allScopes.has('offline.access')) return []

    const highImpactFound = Array.from(allScopes).filter((s) => s !== 'offline_access' && isHighImpact(s))
    if (highImpactFound.length === 0) return []

    const userCount = new Set(
      (sp.oauth2PermissionGrants || [])
        .filter((g) => g.principalId)
        .map((g) => g.principalId as string)
    ).size

    return [
      {
        findingType: 'offline_access_risk',
        severity: 'high',
        title: 'Long-lived token risk via offline_access',
        description:
          `"${sp.displayName}" has offline_access combined with ${highImpactFound.length} ` +
          `high-impact scope(s). This allows the app to access data indefinitely via refresh tokens, ` +
          `even after the user's session ends.`,
        servicePrincipalId: sp.objectId,
        servicePrincipalName: sp.displayName,
        affectedScopes: ['offline_access', ...highImpactFound],
        affectedUserCount: userCount,
        recommendation: this.includeRemediation
          ? 'Review whether persistent offline access is necessary. Consider implementing token lifetime policies.'
          : null,
      },
    ]
  }

  private _detectInactivePrivileged(sp: ServicePrincipal): ShadowOAuthFinding[] {
    if (!sp.signInActivity) return []

    const daysSince = daysSinceLastActivity(sp.signInActivity)
    if (daysSince === null || daysSince < this.inactiveThresholdDays) return []

    const appRoles = getAllAppRoleValues(sp)
    const delegatedScopes = getAllDelegatedScopes(sp)
    const highImpactRoles = Array.from(appRoles).filter(isHighImpact)
    const highImpactScopes = Array.from(delegatedScopes).filter(isHighImpact)
    const allHighImpact = [...highImpactRoles, ...highImpactScopes]

    if (allHighImpact.length === 0) return []

    return [
      {
        findingType: 'inactive_privileged',
        severity: 'medium',
        title: 'Inactive app retaining high-impact permissions',
        description:
          `"${sp.displayName}" has not been used for ${daysSince} days ` +
          `but still holds ${allHighImpact.length} high-impact permission(s). ` +
          `Stale high-privilege apps increase attack surface unnecessarily.`,
        servicePrincipalId: sp.objectId,
        servicePrincipalName: sp.displayName,
        affectedScopes: allHighImpact,
        recommendation: this.includeRemediation
          ? 'Disable or remove this application if it is no longer in use. Otherwise, review and reduce permissions to minimum required.'
          : null,
      },
    ]
  }

  private _detectOrphanedPrivileged(sp: ServicePrincipal): ShadowOAuthFinding[] {
    if (hasOwners(sp)) return []

    const appRoles = getAllAppRoleValues(sp)
    const delegatedScopes = getAllDelegatedScopes(sp)
    const allHighImpact = [
      ...Array.from(appRoles).filter(isHighImpact),
      ...Array.from(delegatedScopes).filter(isHighImpact),
    ]

    if (allHighImpact.length === 0) return []

    return [
      {
        findingType: 'orphaned_privileged',
        severity: 'high',
        title: 'Orphaned app with high-impact permissions',
        description:
          `"${sp.displayName}" has no defined owners but holds ${allHighImpact.length} ` +
          `high-impact permission(s). Without owners, there is no clear accountability for ` +
          `this application's access.`,
        servicePrincipalId: sp.objectId,
        servicePrincipalName: sp.displayName,
        affectedScopes: allHighImpact,
        recommendation: this.includeRemediation
          ? 'Assign at least one owner to this application. If the owning team is unknown, escalate to your security team for review.'
          : null,
      },
    ]
  }

  private _detectUnverifiedPublisherHighImpact(sp: ServicePrincipal): ShadowOAuthFinding[] {
    if (spHasVerifiedPublisher(sp)) return []
    if (!isExternal(sp)) return []

    const appRoles = getAllAppRoleValues(sp)
    const delegatedScopes = getAllDelegatedScopes(sp)
    const allHighImpact = [
      ...Array.from(appRoles).filter(isHighImpact),
      ...Array.from(delegatedScopes).filter(isHighImpact),
    ]

    if (allHighImpact.length === 0) return []

    return [
      {
        findingType: 'unverified_publisher_high_impact',
        severity: 'high',
        title: 'Unverified publisher with high-impact permissions',
        description:
          `"${sp.displayName}" is published by an unverified publisher and has ` +
          `${allHighImpact.length} high-impact permission(s). Unverified publishers ` +
          `have not been validated by Microsoft.`,
        servicePrincipalId: sp.objectId,
        servicePrincipalName: sp.displayName,
        affectedScopes: allHighImpact,
        recommendation: this.includeRemediation
          ? 'Verify the publisher\'s identity through other means, or replace with a verified alternative. Consider revoking access until verification is complete.'
          : null,
      },
    ]
  }
}

// Singleton
export const shadowDetector = new ShadowOAuthDetector()
